### PR TITLE
IPL Mixin for pseudo-labeling

### PR DESCRIPTION
**Pull request by nune-tadevosyan**
# What does this PR do ?

Adding IPLMixin for all ASR models, to be able to run IPL training from the SDP future implementation.
Two scripts that will allow pseudo-label writing and inference.

# Changelog

1. Implementation of IPLMixin
2. Inheritance for all ASR models

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: scripts/pseudo_labeling/write_transcribed_files.py
Comment (nithinraok) [inline scripts/pseudo_labeling/write_transcribed_files.py:1]: if these are handled through SDP, do we need these scripts here?

File: nemo/collections/asr/parts/mixins/mixins.py [hunk 1]
```diff
@@ -895,3 +895,21 @@ def diarize(self, paths2audio_files: List[str], batch_size: int = 1) -> List[str
             Speaker labels
         """
         pass
+
+
+class IPLMixin:
```
Comment (nithinraok) [inline nemo/collections/asr/parts/mixins/mixins.py:900]: instead of this. Add another callbacks class with same functionality here: https://github.com/NVIDIA/NeMo/tree/main/nemo/utils/callbacks

File: scripts/pseudo_labeling/write_transcribed_files.py
Comment (nune-tadevosyan) [reply scripts/pseudo_labeling/write_transcribed_files.py:1]: These are scripts that are run during training on cluster. With SDP we only generate commands and submit them with NeMo-Run.

File: examples/asr/asr_transducer/speech_to_text_rnnt_bpe.py [hunk 2]
```diff
@@ -73,7 +74,11 @@
 def main(cfg):
     logging.info(f'Hydra config: {OmegaConf.to_yaml(cfg)}')
 
-    trainer = pl.Trainer(**resolve_trainer_cfg(cfg.trainer))
+    trainer = pl.Trainer(
+        callbacks=[IPLEpochStopper(enable_stop=True)],
+        **resolve_trainer_cfg(cfg.trainer),
+    )
+
```
Comment (nithinraok) [inline examples/asr/asr_transducer/speech_to_text_rnnt_bpe.py:81]: Remove this

File: nemo/collections/asr/models/asr_model.py [hunk 3]
```diff
@@ -30,7 +31,7 @@
 __all__ = ['ASRModel']
 
 
-class ASRModel(ModelPT, ABC):
+class ASRModel(ModelPT, ABC, IPLMixin):
```
Comment (nithinraok) [inline nemo/collections/asr/models/asr_model.py:34]: remove this

File: nemo/utils/callbacks/ipl_epoch_stop.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        super().__init__()
+        self.enable_stop = bool(enable_stop)
+
+    def on_train_epoch_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+        """
+        Sets `should_stop` stop flag to terminate the training.
+        """
+        super().__init__()
+        if self.enable_stop:
+            trainer.should_stop = True
```
Comment (nithinraok) [inline nemo/utils/callbacks/ipl_epoch_stop.py:39]: Thanks, move this to https://github.com/NVIDIA/NeMo/tree/main/nemo/collections/common/callbacks (lot of similar folders :D , this seemed to be best one for this usecase)

Then add this through exp_manager. See: https://github.com/NVIDIA/NeMo/blob/6e7b414dfcd4ec955278755c24062e66f5ac848e/nemo/utils/exp_manager.py#L708 and you can add similarly for this functionality. No need to add to each class.

**Review by nithinraok — CHANGES_REQUESTED** [review]
